clients/unshaded/src/main/java/tachyon/client/file/FileOutStream.java

```diff
@@ -76,22 +78,25 @@
    */
   public FileOutStream(long fileId, OutStreamOptions options) throws IOException {
     mFileId = fileId;
+    mNonce = ClientContext.getRandomNonNegativeLong();
     mBlockSize = options.getBlockSize();
     mTachyonStorageType = options.getTachyonStorageType();
     mUnderStorageType = options.getUnderStorageType();
     mContext = FileSystemContext.INSTANCE;
     mPreviousBlockOutStreams = new LinkedList<BufferedBlockOutStream>();
     if (mUnderStorageType.isPersist()) {
-      mWorkerClient = BlockStoreContext.INSTANCE.acquireWorkerClient();
-      String sessionUnderStorageFolder = mWorkerClient.getSessionUfsTempFolder();
-      mUnderStorageFile = PathUtils.concatPath(sessionUnderStorageFolder, mFileId);
-      UnderFileSystem underStorageClient =
-          UnderFileSystem.get(mUnderStorageFile, ClientContext.getConf());
-      underStorageClient.mkdirs(sessionUnderStorageFolder, true);
-      mUnderStorageOutputStream = underStorageClient.create(mUnderStorageFile, (int) mBlockSize);
+      FileInfo fileInfo = getFileInfo(mFileId);
+      String ufsPath = fileInfo.getUfsPath();
+      String fileName = PathUtils.temporaryFileName(fileId, mNonce, ufsPath);
```

Comment:
It would save us one RPC at the cost of increasing the size of the window for a potential race. The race is between the location of the temporary file changing after the path is computed and the file is renamed. Now that I think of it, we could do what you suggest, an only regenerate the path if it is invalid. That probably makes the most sense.